interface/wx/webrequest.h

```diff
@@ -0,0 +1,611 @@
+    */
+    wxWebSession();
+
+    /**
+        Create a new request for the specified URL
+
+        @param url
+            The URL of the HTTP resource for this request
+        @param id
+            Optional id sent with events
+    */
+    wxWebRequest* CreateRequest(const wxString& url, int id = wxID_ANY);
+
+    /**
+        Sets a request header in every wxWebRequest created from this session
+        after is has been set.
+
+        A good example for a session-wide request header is the @c User-Agent
+        header.
+
+        @param name Name of the header
+        @param value String value of the header
+    */
+    void SetHeader(const wxString& name, const wxString& value);
+
+    /**
+        Override the default temporary directory that may be used by the
+        session implemention, when required.
+    */
+    void SetTempDir(const wxString& name);
+
+    /**
+        Returns the current temporary directory.
+
+        @see SetTempDir()
+    */
+    const wxString& GetTempDir() const;
+
+    /**
+        Returns the default session
+    */
+    static wxWebSession& GetDefault();
+
+    /**
+        Creates a new wxWebSession object.
+
+        @param backend
+            The backend web session implementation to use.
+
+        @return
+            The created wxWebSession
+    */
+    static wxWebSession* New(const wxString& backend = wxWebSessionBackendDefault);
+
+    /**
+        Allows the registering of new backend for wxWebSession.
+
+        backend can be used as an argument to New().
+
+        @param backend The name for the new backend to be registered under
+        @param factory A shared pointer to the factory which creates the appropriate backend.
+    */
+    static void RegisterFactory(const wxString& backend, wxSharedPtr<wxWebSessionFactory> factory);
+
+    /**
+        Allows to check if the specified backend is available at runtime.
+
+        Usually the default backend should always be available, but e.g. macOS
+        before 10.9 does not have the @c NSURLSession implementation available.
+    */
+    static bool IsBackendAvailable(const wxString& backend);
+};
+
+/**
+    @class wxWebSessionFactory
+
+    An abstract factory class for creation wxWebSession backends.
+
+    Each implementation of wxWebSession should have its own factory.
+
+    @since 3.1.2
+
+    @library{wxnet}
+    @category{net}
+
+    @see wxWebSession
+*/
+class wxWebSessionFactory
+{
+public:
+    /**
+        Creates a new web session object
+    */
+    virtual wxWebSession* Create() = 0;
+};
+
+/**
+    @class wxWebRequestEvent
+
+    A web request event sent during or after server communication.
+
+    @since 3.1.2
+
+    @library{wxnet}
+    @category{net}
+
+    @see wxWebRequest
+*/
+class wxWebRequestEvent : public wxEvent
+{
+public:
+    wxWebRequestEvent();
+    wxWebRequestEvent(wxEventType type, int id, wxWebRequest::State state,
+        wxWebResponse* response = NULL, const wxString& errorDesc = "");
```

Comment:
Is using `""` correct (also in headers)? Other headers use `wxEmptyString` or `wxString()`.

Reply:
Indeed, there are only very few places where `""` was used, and those look more like oversights. At least for consistency, this should probably not be introduced. IIRC the (slightly, but still) more efficient one is `wxString()`.

Reply:
Using `= wxString()` would be better but this can be changed later, so for now I'm resolving this one.